policies for CS batch operations

Monneh opened the pull request.

Monneh committed:
Add policies for CS batch operations

Monneh committed:
Made changes plus done documentation

jashan-lefty commented on inputs/gcp/cloud_storage_batch_operations/google_storage_batch_operations_job/disallow_permanent_object_deletion/c.tf:
```diff
@@ -0,0 +1,18 @@
+resource "google_storage_batch_operations_job" "c" {
+  provider = google-beta
+  project  = "test-project"
+  job_id   = "c"
```
The job_id fied requires two characters to plan successfully you can use c1 for complaints

similarly use nc1 for non-complaints

Follow this for every policy

jashan-lefty commented on policies/gcp/cloud_storage_batch_operations/google_storage_batch_operations_job/rewrite_requires_cmek/policy.rego:
The policy usage here is incorrect

jashan-lefty commented on policies/gcp/cloud_storage_batch_operations/google_storage_batch_operations_job/require_scope_prefix_or_manifest/policy.rego:
instead of creating two conditions blocks just add it in one

values": [null,""]

jashan-lefty reviewed (changes requested):
The job_id fied requires two characters to plan successfully you can use c1 for complaints

similarly use nc1 for non-complaints

Follow this for every policy

Also fix minor changes in your policies as well.

Monneh committed:
feedback addressed

Monneh commented:
Hey @jashan-lefty , I've addressed all the feedback that I got , please do let me know if further changes are needed. thank you

jashan-lefty commented on policies/gcp/cloud_storage_batch_operations/google_storage_batch_operations_job/rewrite_requires_cmek/policy.rego:
```diff
@@ -12,10 +12,10 @@ conditions := [[
         "attribute_path": ["rewrite_object", 0, "kms_key"],
         "policy_type": "pattern whitelist",
-        "values": ["projects/*/locations/*/keyRings/*/cryptoKeys/*", [["*"]]]
+        "values": ["^projects/[^/]+/locations/[^/]+/keyRings/[^/]+/cryptoKeys/[^/]+$"]
```
Hi mate, the previous version of the pattern whitelist was still close to what we do. SO i just wanted you to add the required keys in the array for example:
"values": ["projects/*/locations/*/keyRings/*/cryptoKeys/*", [[""my-project"],["example"],["key"]
]

jashan-lefty replied:
Still it is not correct. Please look at pattern whitelist demo if required or drop the policy

jashan-lefty commented on inputs/gcp/cloud_storage_batch_operations/google_storage_batch_operations_job/disallow_permanent_object_deletion/c.tf:
```diff
@@ -1,7 +1,7 @@
 resource "google_storage_batch_operations_job" "c" {
   provider = google-beta
   project  = "test-project"
-  job_id   = "c"
```
please make sure that you include config files for the policy folders

jashan-lefty replied:
Make sure that you have config.tf in your PR

Monneh replied:
> Make sure that you have config.tf in your PR

Hey bro @jashan-lefty , i have them included in all my policies im not sure sure which one is it

jashan-lefty commented:
looks good

if you change pattern whitelist usage, include the config files. The branch will be good to merge. Thanks

Monneh committed:
Feedback addressed: Updated config files for pattern whitelist usage in rewrite_requires_cmek policy

Monneh commented:
> looks good
> 
> if you change pattern whitelist usage, include the config files. The branch will be good to merge. Thanks

thank You brotha , i fixed it @jashan-lefty

jashan-lefty reviewed (commented):
Please make some small changes and it will be good to merge

Monneh committed:
Fix pattern whitelist implementation in rewrite_requires_cmek policy - now correctly validates CMEK key format with proper wildcard pattern and allowed values array

Monneh committed:
Convert rewrite_requires_cmek to blacklist policy - blocks obviously invalid CMEK key patterns instead of complex pattern whitelist

Monneh committed:
Changed policy approach and ensured config files are included - converted rewrite_requires_cmek from pattern whitelist to blacklist policy, verified all config.tf files exist and are committed

Monneh commented:
@jashan-lefty hey bro I've switched the pattern whitelist issue that I had , I've changed it to pattern blacklist and done it . for the other one where you mentioned to add config.tf I've already had them but just incase ill upload screenshots that my policies run locally ?

Monneh committed:
Rework cloud storage batch operation policies - all working as intended

- Fixed Terraform configurations to use hardcoded project values
- Corrected non-compliant (nc) resources to demonstrate proper policy violations
- All 4 policies now correctly detect and flag security violations:
  * forbid_unsetting_object_holds: flags UNSET event_based_hold
  * disallow_permanent_object_deletion: flags permanent deletion enabled
  * require_scope_prefix_or_manifest: flags unscoped operations
  * rewrite_requires_cmek: flags invalid CMEK key paths
- OPA evaluations working correctly with proper non-compliant resource detection

Monneh commented:
<img width="1662" height="357" alt="image" src="https://github.com/user-attachments/assets/e2dac971-1c5f-43d8-9e29-18fc49c8ca90" />
<img width="1674" height="775" alt="image" src="https://github.com/user-attachments/assets/39f50574-64f7-473b-b8e6-9c880592ebfd" />
<img width="726" height="679" alt="image" src="https://github.com/user-attachments/assets/059d85a7-8aab-4e56-b525-6bea545e7639" />
<img width="1772" height="316" alt="image" src="https://github.com/user-attachments/assets/05694e7f-20c4-43d7-a47b-4fc007fb1751" />
<img width="1765" height="627" alt="image" src="https://github.com/user-attachments/assets/43b75b9f-a1bb-407f-90eb-8d69e24da6b1" />
